src/Psl/Type/Internal/ShapeType.php

```diff
@@ -33,6 +40,47 @@ public function __construct(
      * @return array<Tk, Tv>
      */
     public function coerce(mixed $value): array
+    {
+        // To whom reads this: yes, I hate this stuff as passionately as you do :-)
+        if (! \is_array($value)) {
+            // Fallback to slow implementation - unhappy path
+            return $this->coerceIterable($value);
+        }
+
+        if (\array_keys(\array_intersect_key($value, $this->requiredElements)) !== \array_keys($this->requiredElements)) {
+            // Fallback to slow implementation - unhappy path
+            return $this->coerceIterable($value);
+        }
+
+        if (! $this->allow_unknown_fields && \array_keys($value) !== \array_keys($this->elements_types)) {
+            // Fallback to slow implementation - unhappy path
+            return $this->coerceIterable($value);
+        }
+
+        $coerced = [];
+
+        try {
+            foreach (\array_intersect_key($this->elements_types, $value) as $key => $type) {
+                $coerced[$key] = $type->coerce($value[$key]);
+            }
+        } catch (CoercionException) {
+            // Fallback to slow implementation - unhappy path. Prevents having to eagerly compute traces.
+            $this->coerceIterable($value);
+        }
+
+        foreach (\array_diff_key($value, $this->elements_types) as $key => $additionalValue) {
+            $coerced[$key] = $additionalValue;
+        }
+
+        return $coerced;
+    }
+
+    /**
+     * @throws CoercionException
+     *
+     * @return array<Tk, Tv>
+     */
+    private function coerceIterable(mixed $value): array
     {
         /** @psalm-suppress MissingThrowsDocblock */
         if (Type\iterable(Type\mixed(), Type\mixed())->matches($value)) {
```

Comment:
Couldn't this just check `is_iterable`?

Reply:
Possibly, but I decided to not touch this code branch for now.

Reply:
Understood, just wanted to point to it just in case.

Reply:
I've adjusted this path - more optimizations also possible, but `iterable` + `ShapeType` is really not my focus for now :) 

Benchmarks should allow for future iterations